File: pkg/sdk/create.go
package sdk

import (
	"context"
	"encoding/json"
	"runtime"

	"github.com/jingkaihe/matchlock/internal/errx"
	"github.com/jingkaihe/matchlock/pkg/api"
)

func (c *Client) Create(opts CreateOptions) (string, error) {
	if opts.Image == "" {
		return "", ErrImageRequired
	}
	if opts.CPUs == 0 {
		opts.CPUs = api.DefaultCPUs
	}
	vcpus, ok := api.VCPUCount(opts.CPUs)
	if !ok {
		return "", ErrInvalidCPUCount
	}
	if vcpus > runtime.NumCPU() {
		return "", ErrInvalidCPUCount
	}
	if opts.MemoryMB == 0 {
		opts.MemoryMB = api.DefaultMemoryMB
	}
	if opts.DiskSizeMB == 0 {
		opts.DiskSizeMB = api.DefaultDiskSizeMB
	}
	if opts.TimeoutSeconds == 0 {
		opts.TimeoutSeconds = api.DefaultTimeoutSeconds
	}
	if opts.NetworkMTU < 0 {
		return "", ErrInvalidNetworkMTU
	}
	if opts.NoNetwork && (len(opts.AllowedHosts) > 0 || len(opts.Secrets) > 0 || opts.ForceInterception || opts.NetworkInterception != nil) {
		return "", ErrNoNetworkConflict
	}
	for _, mapping := range opts.AddHosts {
		if err := api.ValidateAddHost(mapping); err != nil {
			return "", errx.Wrap(ErrInvalidAddHost, err)
		}
	}

	wireVFS, localHooks, localMutateHooks, localActionHooks, err := compileVFSHooks(opts.VFSInterception)
	if err != nil {
		return "", err
	}
	wireNetwork, localNetworkHooks, err := compileNetworkHooks(opts.NetworkInterception)
	if err != nil {
		return "", err
	}

	c.stopNetworkHookServer()
	startedNetworkHookServer := false
	if len(localNetworkHooks) > 0 {
		socketPath, listener, tempDir, err := startNetworkHookServer(localNetworkHooks)
		if err != nil {
			return "", errx.With(ErrInvalidNetworkHook, ": start callback server: %v", err)
		}
		if wireNetwork == nil {
			wireNetwork = &api.NetworkInterceptionConfig{}
		}
		wireNetwork.CallbackSocket = socketPath
		c.networkHookMu.Lock()
		c.networkHooks = localNetworkHooks
		c.networkHookSocket = socketPath
		c.networkHookListener = listener
		c.networkHookTempDir = tempDir
		c.networkHookMu.Unlock()
		startedNetworkHookServer = true
	}

	params := map[string]interface{}{
		"image": opts.Image,
		"resources": map[string]interface{}{
			"cpus":            opts.CPUs,
			"memory_mb":       opts.MemoryMB,
			"disk_size_mb":    opts.DiskSizeMB,
			"timeout_seconds": opts.TimeoutSeconds,
		},
	}
	if opts.KernelRef != "" {
		params["kernel"] = map[string]interface{}{"ref": opts.KernelRef}
	}

	if opts.Privileged {
		params["privileged"] = true
	}

	if network := buildCreateNetworkParams(opts, wireNetwork); network != nil {
		params["network"] = network
	}

	if len(opts.Mounts) > 0 || opts.Workspace != "" || wireVFS != nil {
		vfs := make(map[string]interface{})
		if len(opts.Mounts) > 0 {
			vfs["mounts"] = opts.Mounts
		}
		if opts.Workspace != "" {
			vfs["workspace"] = opts.Workspace
		}
		if wireVFS != nil {
			vfs["interception"] = wireVFS
		}
		params["vfs"] = vfs
	}

	if len(opts.Env) > 0 {
		params["env"] = opts.Env
	}

	if opts.ImageConfig != nil {
		params["image_config"] = opts.ImageConfig
	}
	if opts.LaunchEntrypoint {
		params["launch_entrypoint"] = true
	}

	result, err := c.sendRequest("create", params)
	if err != nil {
		if startedNetworkHookServer {
			c.stopNetworkHookServer()
		}
		return "", err
	}

	var createResult struct {
		ID string `json:"id"`
	}
	if err := json.Unmarshal(result, &createResult); err != nil {
		if startedNetworkHookServer {
			c.stopNetworkHookServer()
		}
		return "", errx.Wrap(ErrParseCreateResult, err)
	}

	c.vmID = createResult.ID
	c.setVFSHooks(localHooks, localMutateHooks, localActionHooks)

	if len(opts.PortForwards) > 0 {
		if _, err := c.portForwardMappings(context.Background(), opts.PortForwardAddresses, opts.PortForwards); err != nil {
			return c.vmID, err
		}
	}
	return c.vmID, nil
}

func buildCreateNetworkParams(opts CreateOptions, wireInterception *api.NetworkInterceptionConfig) map[string]interface{} {
	hasAllowedHosts := len(opts.AllowedHosts) > 0
	hasAddHosts := len(opts.AddHosts) > 0
	hasSecrets := len(opts.Secrets) > 0
	hasDNSServers := len(opts.DNSServers) > 0
	hasHostname := len(opts.Hostname) > 0
	hasMTU := opts.NetworkMTU > 0
	hasNoNetwork := opts.NoNetwork
	hasForceInterception := opts.ForceInterception
	hasNetworkInterception := wireInterception != nil
	blockPrivateIPs, hasBlockPrivateIPsOverride := resolveCreateBlockPrivateIPs(opts)

	includeNetwork := hasAllowedHosts || hasAddHosts || hasSecrets || hasDNSServers || hasHostname || hasMTU || hasNoNetwork || hasBlockPrivateIPsOverride || hasForceInterception || hasNetworkInterception
	if !includeNetwork {
		return nil
	}

	if hasNoNetwork {
		network := map[string]interface{}{
			"no_network": true,
		}
		if hasAddHosts {
			network["add_hosts"] = opts.AddHosts
		}
		if hasDNSServers {
			network["dns_servers"] = opts.DNSServers
		}
		if hasHostname {
			network["hostname"] = opts.Hostname
		}
		return network
	}

	// Create config merges replace network defaults wholesale. Preserve default
	// private-IP blocking unless explicitly overridden.
	if !hasBlockPrivateIPsOverride {
		blockPrivateIPs = true
	}

	network := map[string]interface{}{
		"allowed_hosts":     opts.AllowedHosts,
		"block_private_ips": blockPrivateIPs,
	}
	if hasForceInterception || hasNetworkInterception {
		network["intercept"] = true
	}
	if hasNetworkInterception {
		network["interception"] = wireInterception
	}
	if hasAddHosts {
		network["add_hosts"] = opts.AddHosts
	}
	if hasSecrets {
		secrets := make(map[string]interface{})
		for _, s := range opts.Secrets {
			secret := map[string]interface{}{
				"value": s.Value,
				"hosts": s.Hosts,
			}
			if s.Placeholder != "" {
				secret["placeholder"] = s.Placeholder
			}
			secrets[s.Name] = secret
		}
		network["secrets"] = secrets
	}
	if hasDNSServers {
		network["dns_servers"] = opts.DNSServers
	}
	if hasHostname {
		network["hostname"] = opts.Hostname
	}
	if hasMTU {
		network["mtu"] = opts.NetworkMTU
	}
	return network
}

func resolveCreateBlockPrivateIPs(opts CreateOptions) (value bool, hasOverride bool) {
	if opts.BlockPrivateIPsSet {
		return opts.BlockPrivateIPs, true
	}
	// Backward compatibility: old callers could only express explicit true.
	if opts.BlockPrivateIPs {
		return true, true
	}
	return false, false
}

File: sdk/typescript/src/client/create.ts
import { MatchlockError } from "../errors";
import type { CreateOptions } from "../types";
import {
  DEFAULT_CPUS,
  DEFAULT_DISK_SIZE_MB,
  DEFAULT_MEMORY_MB,
  DEFAULT_TIMEOUT_SECONDS,
} from "./utils";
import type {
  JSONObject,
  JSONValue,
  WireNetworkInterceptionConfig,
  WireVFSInterceptionConfig,
} from "./wire";

export function validateCreateOptions(options: CreateOptions): void {
  if (!options.image) {
    throw new MatchlockError("image is required (e.g., alpine:latest)");
  }
  if ((options.networkMtu ?? 0) < 0) {
    throw new MatchlockError("network mtu must be > 0");
  }
  if (
    options.noNetwork &&
    ((options.allowedHosts?.length ?? 0) > 0 ||
      (options.secrets?.length ?? 0) > 0 ||
      options.forceInterception === true ||
      options.networkInterception !== undefined)
  ) {
    throw new MatchlockError(
      "no network cannot be combined with allowed hosts, secrets, forced interception, or network interception rules",
    );
  }
}

export function buildCreateParams(
  options: CreateOptions,
  wireVFS: WireVFSInterceptionConfig | undefined,
  wireNetworkInterception: WireNetworkInterceptionConfig | undefined,
): JSONObject {
  const resources = {
    cpus: options.cpus ?? DEFAULT_CPUS,
    memory_mb: options.memoryMb || DEFAULT_MEMORY_MB,
    disk_size_mb: options.diskSizeMb || DEFAULT_DISK_SIZE_MB,
    timeout_seconds: options.timeoutSeconds || DEFAULT_TIMEOUT_SECONDS,
  };

  const params: JSONObject = {
    image: options.image ?? "",
    resources,
  };

  if (options.kernelRef) {
    params.kernel = { ref: options.kernelRef };
  }

  if (options.privileged) {
    params.privileged = true;
  }

  const network = buildCreateNetworkParams(options, wireNetworkInterception);
  if (network) {
    params.network = network;
  }

  if (
    (options.mounts && Object.keys(options.mounts).length > 0) ||
    options.workspace ||
    wireVFS
  ) {
    const vfs: JSONObject = {};
    if (options.mounts && Object.keys(options.mounts).length > 0) {
      const mounts: JSONObject = {};
      for (const [guestPath, config] of Object.entries(options.mounts)) {
        const mount: JSONObject = {
          type: config.type ?? "memory",
        };
        if (config.hostPath) {
          mount.host_path = config.hostPath;
        }
        if (config.readonly) {
          mount.readonly = true;
        }
        if (config.ownerUID !== undefined) {
          mount.owner_uid = config.ownerUID;
        }
        if (config.ownerGID !== undefined) {
          mount.owner_gid = config.ownerGID;
        }
        mounts[guestPath] = mount;
      }
      vfs.mounts = mounts;
    }
    if (options.workspace) {
      vfs.workspace = options.workspace;
    }
    if (wireVFS) {
      vfs.interception = wireVFS as unknown as JSONValue;
    }
    params.vfs = vfs;
  }

  if (options.env && Object.keys(options.env).length > 0) {
    params.env = options.env;
  }

  if (options.imageConfig) {
    const imageConfig: JSONObject = {};
    if (options.imageConfig.user) {
      imageConfig.user = options.imageConfig.user;
    }
    if (options.imageConfig.workingDir) {
      imageConfig.working_dir = options.imageConfig.workingDir;
    }
    if (options.imageConfig.entrypoint) {
      imageConfig.entrypoint = [...options.imageConfig.entrypoint];
    }
    if (options.imageConfig.cmd) {
      imageConfig.cmd = [...options.imageConfig.cmd];
    }
    if (options.imageConfig.env) {
      imageConfig.env = { ...options.imageConfig.env };
    }
    params.image_config = imageConfig;
  }
  if (options.launchEntrypoint) {
    params.launch_entrypoint = true;
  }

  return params;
}

function resolveCreateBlockPrivateIPs(opts: CreateOptions): {
  value: boolean;
  hasOverride: boolean;
} {
  if (opts.blockPrivateIPsSet) {
    return { value: !!opts.blockPrivateIPs, hasOverride: true };
  }
  if (opts.blockPrivateIPs) {
    return { value: true, hasOverride: true };
  }
  return { value: false, hasOverride: false };
}

function buildCreateNetworkParams(
  opts: CreateOptions,
  wireInterception: WireNetworkInterceptionConfig | undefined,
): JSONObject | undefined {
  const hasAllowedHosts = (opts.allowedHosts?.length ?? 0) > 0;
  const hasAddHosts = (opts.addHosts?.length ?? 0) > 0;
  const hasSecrets = (opts.secrets?.length ?? 0) > 0;
  const hasDNSServers = (opts.dnsServers?.length ?? 0) > 0;
  const hasHostname = (opts.hostname?.length ?? 0) > 0;
  const hasMTU = (opts.networkMtu ?? 0) > 0;
  const hasNoNetwork = opts.noNetwork === true;
  const hasForceInterception = opts.forceInterception === true;
  const hasNetworkInterception = wireInterception !== undefined;

  const blockPrivate = resolveCreateBlockPrivateIPs(opts);

  const includeNetwork =
    hasAllowedHosts ||
    hasAddHosts ||
    hasSecrets ||
    hasDNSServers ||
    hasHostname ||
    hasMTU ||
    hasNoNetwork ||
    blockPrivate.hasOverride ||
    hasForceInterception ||
    hasNetworkInterception;

  if (!includeNetwork) {
    return undefined;
  }

  if (hasNoNetwork) {
    const network: JSONObject = {
      no_network: true,
    };
    if (hasAddHosts) {
      network.add_hosts = (opts.addHosts ?? []).map((mapping) => ({
        host: mapping.host,
        ip: mapping.ip,
      }));
    }
    if (hasDNSServers) {
      network.dns_servers = opts.dnsServers ?? [];
    }
    if (hasHostname) {
      network.hostname = opts.hostname ?? "";
    }
    return network;
  }

  const network: JSONObject = {
    allowed_hosts: opts.allowedHosts ?? [],
    block_private_ips: blockPrivate.hasOverride ? blockPrivate.value : true,
  };
  if (hasForceInterception || hasNetworkInterception) {
    network.intercept = true;
  }
  if (hasNetworkInterception) {
    network.interception = wireInterception as unknown as JSONValue;
  }

  if (hasAddHosts) {
    network.add_hosts = (opts.addHosts ?? []).map((mapping) => ({
      host: mapping.host,
      ip: mapping.ip,
    }));
  }

  if (hasSecrets) {
    const secrets: JSONObject = {};
    for (const secret of opts.secrets ?? []) {
      const wireSecret: JSONObject = {
        value: secret.value,
        hosts: secret.hosts ?? [],
      };
      if (secret.placeholder) {
        wireSecret.placeholder = secret.placeholder;
      }
      secrets[secret.name] = wireSecret;
    }
    network.secrets = secrets;
  }

  if (hasDNSServers) {
    network.dns_servers = opts.dnsServers ?? [];
  }

  if (hasHostname) {
    network.hostname = opts.hostname ?? "";
  }

  if (hasMTU) {
    network.mtu = opts.networkMtu ?? 0;
  }

  return network;
}
